### ISSUE TITLE
Different Output from Libtorch c++than pytorch

### ISSUE THREAD

**Issue by Arki99** [pytorch/pytorch#42987]
I followed [https://pytorch.org/tutorials/advanced/cpp_export.html](url) to trace a model from existing weights in python as shown below

`        

        if self.conf.use_cuda:
            torch.set_default_tensor_type('torch.cuda.FloatTensor')
        else:
            torch.set_default_tensor_type('torch.FloatTensor')

        self.learner = face_learner(self.conf, self.device)
        # load pretrained model for learner object
        self.learner.model.load_state_dict(torch.load(self.conf.learner_model_path))
        self.learner.model.eval()

        # read the example image used for tracing
        image=cv2.imread("videos/example.jpg",1)

        test_transform = trans.Compose([
                    trans.ToTensor(),
                    trans.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
                ])

        resized_image = cv2.resize(image, (112, 112))
        # Use torch.jit.trace to generate a torch.jit.ScriptModule via tracing.
        traced_script_module = torch.jit.trace(self.learner.model, test_transform(resized_image).unsqueeze(0).to(self.device))
        traced_script_module.save("traced_facelearner_model_new.pt")
`

The face learner model gives 1x512 embedding as output.

`

                torch::jit::script::Module model = torch::jit::load("traced_facelearner_model_new.pt");
		model.to(torch::kCUDA);
		model.eval();

		cv::Mat visibleFrame = cv::imread("example.jpg");

		cv::resize(visibleFrame, visibleFrame, cv::Size(112, 112));
		at::Tensor tensor_image = torch::from_blob(visibleFrame.data, { 1, visibleFrame.rows, visibleFrame.cols, 3 }, at::kByte);
		tensor_image = tensor_image.permute({ 0, 3, 1, 2 });
		tensor_image = tensor_image.to(at::kFloat);

		tensor_image[0][0] = tensor_image[0][0].sub(0.5).div(0.5);
		tensor_image[0][1] = tensor_image[0][1].sub(0.5).div(0.5);
		tensor_image[0][2] = tensor_image[0][2].sub(0.5).div(0.5);

		tensor_image = tensor_image.to(torch::kCUDA);
		std::vector<torch::jit::IValue> input;
		input.emplace_back(tensor_image);
		// Execute the model and turn its output into a tensor.
		auto output = model.forward(input).toTensor();
`

I'm pretty sure my preprocessing steps on the image in libtorch and pytorch are the same. But the output embedding is still different. 
for eg: python output of first 3 values in the embedding
tensor([[ 2.3617e-02, -1.3115e-02,  7.1695e-02,

c++ output of first 3 values
 Columns 1 to 10 -0.0248 -0.0245 -0.0204

**Comment by glaringlee** [comment]
@Arki99 
Several questions:
1. Are you using the same version of pytorch and libtorch?
2. You said the preprocessing steps are the same, you mean before calling forward, the values within all the inputs between libtorch and pytorch are all the same, correct?
3. Are you able to share more code? it would be great if you can share me some code that can reproduce your issue.

Thanks a lot.

**Comment by Arki99** [comment]
1. pytorch 1.4 and libtorch 1.5
2. Yes, that's correct
3. I'm sorry I can only share this much.

**Comment by glaringlee** [comment]
@Arki99 
ok, can you try using latest pytorch and latest libtorch? We have a bunch of fixes between 1.4 and 1.5.
If you can not provide code, can you pls provide a Tensor (with values) or something so we can reproduce the error?

**Comment by Arki99** [comment]
@glaringlee 
I tried using pytorch 1.5 and libtorch 1.5, the issue still persists.

You can find the c++, python codes and the model below

[https://drive.google.com/drive/folders/1wFKbIqoValQRe_UICe4IxfEwLcoEwUca?usp=sharing](url)

**Comment by glaringlee** [comment]
@Arki99 Sry, I might not clear enough. Can you try the latest pytorch and libtorch?

The inputs for both python and c++ are the same, and c++ code call the same underlying as python. Subscribe jit team as well to check whether this is related to jit. Thx

**Comment by Arki99** [comment]
@glaringlee Tried with the latest pytorch and libtorch. Still facing the same issue.

**Comment by glaringlee** [comment]
@Arki99 
Hi, I was trying to run your code, but I can not load your pt file on my end. Can you tell me what version of pytorch you were using to create that pt file? And what version of libtorch you were use?

And even I can load your pt file and run your module, that's a 100+mb pt file, it is not feasible for me to trace it without knowing your module. Can you provide a simplified module which can reproduce your issue?  And provide me your original implementation, so I can reproduce it on my end. thx

**Comment by Arki99** [comment]
@glaringlee https://stackoverflow.com/questions/63502473/different-output-from-libtorch-c-and-pytorch/63504086#63504086
The answer provided here fixed my problem. 
`

       at::Tensor tensor_image = torch::from_blob(visibleFrame.data, {  visibleFrame.rows, 
                                                    visibleFrame.cols, 3 }, at::kByte);
        
       tensor_image = tensor_image.to(at::kFloat).div(255).unsqueeze(0);`

Adding these in the C++ code is now giving the same output as my python code
